assets/js/googlesitekit/widgets/datastore/widgets.js

```diff
@@ -0,0 +1,258 @@
+		}
+		const settingsToUse = {
+			priority: 10,
+			...settings,
+		};
+		delete settingsToUse.component;
+
+		yield {
+			payload: { slug, settings: settingsToUse },
+			type: REGISTER_WIDGET,
+		};
+
+		return {};
+	},
+};
+
+export const controls = {};
+
+export const reducer = ( state, { type, payload } ) => {
+	switch ( type ) {
+		case ASSIGN_WIDGET: {
+			const { slug, areaSlugs } = payload;
+
+			const { areaWidgets } = state;
+			areaSlugs.forEach( ( areaSlug ) => {
+				if ( areaWidgets[ areaSlug ] === undefined ) {
+					areaWidgets[ areaSlug ] = [];
+				}
+
+				if ( ! areaWidgets[ areaSlug ].includes( slug ) ) {
+					areaWidgets[ areaSlug ].push( slug );
+				}
+			} );
+
+			return {
+				...state,
+				areaWidgets,
+			};
+		}
+
+		case REGISTER_WIDGET: {
+			const { slug, settings } = payload;
+
+			if ( state.widgets[ slug ] !== undefined ) {
+				global.console.warn( `Could not register widget with slug "${ slug }". Widget "${ slug }" is already registered.` );
+
+				return { ...state };
+			}
+
+			return {
+				...state,
+				widgets: {
+					...state.widgets,
+					[ slug ]: { ...settings, slug },
+				},
+			};
+		}
+
+		case SET_WIDGET_COMPONENT_KEY: {
+			const { registryKey } = payload;
+
+			return {
+				...state,
+				registryKey,
+			};
+		}
+
+		default: {
+			return { ...state };
+		}
+	}
+};
+
+export const resolvers = {
+};
+
+export const selectors = {
+	/**
+	 * Checks if a widget has been registered with a given slug.
+	 *
+	 * Returns `true` if the widget area has been registered.
+	 * Returns `false` if the widget area has NOT been registered.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {Object}  state Data store's state.
+	 * @param  {string}  slug  Widget's slug.
+	 * @return {boolean}       `true`/`false` based on whether widget has been registered.
+	 */
+	isWidgetRegistered( state, slug ) {
+		const { widgets } = state;
+
+		return widgets[ slug ] !== undefined;
+	},
+
+	/**
+	 * Returns all widgets registered for a given widget area.
+	 *
+	 * Returns an array of all widgets for a given area.
+	 * The widgets are returned in order of their priority, so can be rendered in
+	 * the order provided by the selector.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {Object} state          Data store's state.
+	 * @param  {string} widgetAreaSlug Widget context to get areas for.
+	 * @return {Array}                 An ordered array of widgets for this area.
+	 */
+	getWidgets: createRegistrySelector( ( select ) => ( state, widgetAreaSlug ) => {
+		invariant( widgetAreaSlug, 'widgetAreaSlug is required.' );
+
+		const { areaWidgets, widgets } = state;
+
+		const registryKey = select( STORE_NAME ).getWidgetRegistryKey();
+
+		return Object.values( widgets ).filter( ( widget ) => {
+			return areaWidgets[ widgetAreaSlug ] && areaWidgets[ widgetAreaSlug ].includes( widget.slug );
+		} ).sort( ( widgetA, widgetB ) => {
+			return ( widgetA.priority >= widgetB.priority ) ? 1 : -1;
+		} ).map( ( widget ) => {
```

Comment:
See other comment around widget areas, may be faster to do the sorting in the register reducer? What do you think?

Reply:
I think it's best to memoise these selectors than storing data sorted, but I think that's a good call to explore how to improve selector performance as a separate issue 👍